### PR TITLE
OE-317

### PR DESCRIPTION
**Pull request by apigram**
Advanced search now correctly searches across systemic diagnoses as well as ophthalmic diagnoses.
Systemic diagnoses are now included in the diagnoses table in Advanced Search. All systemic diagnoses are marked as secondary and also now have the diagnosis origin listed.
Fixed PHPDoc hints for newer relational properties in the Patient class.
Added three new views to easily collate the most up-to-date systemic diagnoses for each patient.

### CODE REVIEW

File: protected/models/Patient.php [hunk 1]
```diff
@@ -57,8 +57,10 @@
  * @property SecondaryDiagnosis[] $secondarydiagnoses
  * @property EthnicGroup $ethnic_group
  * @property CommissioningBody[] $commissioningbodies
- * @property SocialHistory $socialhistory
- * @property TrialPatient[] trials
+ * @property \OEModule\OphCiExamination\models\SocialHistory $socialhistory
+ * @property TrialPatient[] $trials
```
Comment (marshl) [inline protected/models/Patient.php]: Is this the right type? Or maybe the variable has been named incorrectly.

File: protected/models/Patient.php [hunk 2]
```diff
@@ -171,6 +173,7 @@ public function relations()
             'diagnoses' => array(self::MANY_MANY, '\OEModule\OphCiExamination\models\OphCiExamination_Diagnosis', 'patient_diagnosis_assignment(patient_id, id)',
                 'alias' => 'patient_diagnoses',
                 'order' => 'patient_diagnoses.principal DESC', ),
+            'systemic_diagnoses' => array(self::MANY_MANY, '\OEModule\OphCiExamination\models\SystemicDiagnoses_Diagnosis', 'patient_systemic_diagnosis(patient_id, id)'),
```
Comment (marshl) [inline protected/models/Patient.php]: Many to Many, is this right? Would multiple patients have the same patient-diagnosis?

File: protected/models/PatientDiagnosisParameter.php [hunk 3]
```diff
@@ -114,7 +114,17 @@ public function query($searchProvider)
   ON paa.patient_id = p.id
 LEFT JOIN disorder d
   ON d.id = paa.disorder_id
-WHERE LOWER(d.term) LIKE LOWER(:p_d_value_$this->id)";
+WHERE LOWER(d.term) LIKE LOWER(:p_d_value_$this->id)
+
+UNION
+
+SELECT DISTINCT p2.id
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: Same issues as with the unit tests

File: protected/models/PatientDiagnosisParameter.php [hunk 4]
```diff
@@ -114,7 +114,17 @@ public function query($searchProvider)
...
+FROM patient p2
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: As above

File: protected/models/PatientDiagnosisParameter.php [hunk 5]
```diff
@@ -114,7 +114,17 @@ public function query($searchProvider)
...
 LEFT JOIN disorder d
   ON d.id = paa.disorder_id
-WHERE LOWER(d.term) LIKE LOWER(:p_d_value_$this->id)";
+WHERE LOWER(d.term) LIKE LOWER(:p_d_value_$this->id)
+
+UNION
+
+SELECT DISTINCT p2.id
+FROM patient p2 
+LEFT JOIN patient_systemic_diagnosis psd
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: As above

File: protected/models/PatientDiagnosisParameter.php [hunk 6]
```diff
@@ -114,7 +114,17 @@ public function query($searchProvider)
...
+  ON psd.patient_id = p2.id
+LEFT JOIN disorder d2
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: As above

File: protected/models/PatientDiagnosisParameter.php [hunk 7]
```diff
@@ -132,7 +142,27 @@ public function query($searchProvider)
 LEFT JOIN episode ep
   ON ep.id = e.episode_id
 WHERE LOWER(d.term) LIKE LOWER(:p_d_value_$this->id)
-  AND ep.firm_id = :p_d_firm_$this->id";
+  AND ep.firm_id = :p_d_firm_$this->id
+  
+UNION
+
+SELECT DISTINCT p2.id
+FROM patient p2
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: Same issues as commented on in the unit test PR

File: protected/models/PatientDiagnosisParameter.php [hunk 8]
```diff
@@ -132,7 +142,27 @@ public function query($searchProvider)
...
   ON ep.id = e.episode_id
 WHERE LOWER(d.term) LIKE LOWER(:p_d_value_$this->id)
-  AND ep.firm_id = :p_d_firm_$this->id";
+  AND ep.firm_id = :p_d_firm_$this->id
+  
+UNION
+
+SELECT DISTINCT p2.id
+FROM patient p2 
+LEFT JOIN patient_systemic_diagnosis psd
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: As above

File: protected/models/PatientDiagnosisParameter.php [hunk 9]
```diff
@@ -132,7 +142,27 @@ public function query($searchProvider)
...
+  ON psd.patient_id = p2.id
+LEFT JOIN disorder d2
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: As above

File: protected/models/PatientDiagnosisParameter.php [hunk 10]
```diff
@@ -132,7 +142,27 @@ public function query($searchProvider)
 LEFT JOIN episode ep
   ON ep.id = e.episode_id
 WHERE LOWER(d.term) LIKE LOWER(:p_d_value_$this->id)
-  AND ep.firm_id = :p_d_firm_$this->id";
+  AND ep.firm_id = :p_d_firm_$this->id
+  
+UNION
+
+SELECT DISTINCT p2.id
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: DISTINCT and UNION combined are unecessary

File: protected/models/PatientDiagnosisParameter.php [hunk 11]
```diff
@@ -132,7 +142,27 @@ public function query($searchProvider)
...
+FROM patient p2 
+LEFT JOIN patient_systemic_diagnosis psd
+  ON psd.patient_id = p2.id
+LEFT JOIN disorder d2
+  ON d2.id = psd.disorder_id
+LEFT JOIN et_ophciexamination_systemic_diagnoses et_systemic
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: As above

File: protected/models/PatientDiagnosisParameter.php [hunk 12]
```diff
@@ -132,7 +142,27 @@ public function query($searchProvider)
...
+  ON et_systemic.id = psd.element_id
+LEFT JOIN latest_systemic_examination_events latest2
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: As above

File: protected/models/PatientDiagnosisParameter.php [hunk 13]
```diff
@@ -132,7 +142,27 @@ public function query($searchProvider)
...
+  ON latest2.event_id = et_systemic.event_id
+  AND latest2.patient_id = p2.id
+LEFT JOIN event e2
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: As above

File: protected/models/PatientDiagnosisParameter.php [hunk 14]
```diff
@@ -132,7 +142,27 @@ public function query($searchProvider)
...
+  ON e2.id = latest2.event_id
+LEFT JOIN episode ep2
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: As above

File: protected/models/PatientDiagnosisParameter.php [hunk 15]
```diff
@@ -132,7 +142,27 @@ public function query($searchProvider)
...
+  ON ep2.id = e2.episode_id
+WHERE LOWER(d2.term) LIKE LOWER(:p_d_value_$this->id)
+  AND ep2.firm_id = :p_d_firm_$this->id";
```
Comment (marshl) [inline protected/models/PatientDiagnosisParameter.php]: As above

File: protected/modules/OphCiExamination/migrations/m171011_023359_patient_systemic_diagnosis_view.php [hunk 16]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+class m171011_023359_patient_systemic_diagnosis_view extends CDbMigration
+{
+    public function createView($view_name, $select)
+    {
+        $this->dbConnection->createCommand('create view ' . $view_name . ' as ' . $select)->execute();
```
Comment (marshl) [inline protected/modules/OphCiExamination/migrations/m171011_023359_patient_systemic_diagnosis_view.php]: OEMigration might already have this ability

File: protected/modules/OphCiExamination/migrations/m171011_023359_patient_systemic_diagnosis_view.php [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+    public function dropView($view_name)
+    {
+        $this->dbConnection->createCommand('drop view ' . $view_name)->execute();
+    }
+
+    public function up()
+    {
+        $this->createView('systemic_examination_events', <<<EOSQL
+select event_date, event.created_date, event_id, patient_id from et_ophciexamination_diagnoses et
```
Comment (marshl) [inline protected/modules/OphCiExamination/migrations/m171011_023359_patient_systemic_diagnosis_view.php]: Try to use the correct capitlisation

Comment (marshl) [inline protected/modules/OphCiExamination/migrations/m171011_023359_patient_systemic_diagnosis_view.php]: Try to use meaningful aliases

File: protected/modules/OphCiExamination/migrations/m171011_023359_patient_systemic_diagnosis_view.php [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+  latest.patient_id as patient_id,
+  aa.element_id,
+  aa.disorder_id,
+  aa.side_id,
+  aa.secondary_diagnosis_id,
+  aa.last_modified_user_id,
+  aa.last_modified_date,
+  aa.created_user_id,
+  aa.created_date
+from ophciexamination_systemic_diagnoses_diagnosis as aa
```
Comment (marshl) [inline protected/modules/OphCiExamination/migrations/m171011_023359_patient_systemic_diagnosis_view.php]: Try to use meaningful aliases

File: protected/widgets/views/patientDiagnosesAndMedicationsWidget.php [hunk 19]
```diff
@@ -26,6 +26,13 @@
                                     <td><?php echo $diagnosis->element_diagnoses->event ? CHtml::encode(Helper::convertDate2NHS($diagnosis->element_diagnoses->event->event_date)) : 'Unknown'; ?></td>
                                 </tr>
                             <?php endforeach; ?>
+                            <?php foreach ($this->patient->systemic_diagnoses as $diagnosis):?>
```
Comment (marshl) [inline protected/widgets/views/patientDiagnosesAndMedicationsWidget.php]: Try to run autoformat after changing files

File: protected/modules/OphCiExamination/migrations/m171011_023359_patient_systemic_diagnosis_view.php (hunk 16 above)
Comment (apigram) [reply protected/modules/OphCiExamination/migrations/m171011_023359_patient_systemic_diagnosis_view.php]: I don't believe it does. I've seen other migrations that have this function defined in them as well.

File: protected/models/Patient.php (hunk 1 above)
Comment (apigram) [reply protected/models/Patient.php]: Everything in OphCiExamination is namespaced, so I believe it has to be declared in this way.

File: protected/models/Patient.php (hunk 2 above)
Comment (apigram) [reply protected/models/Patient.php]: This is a relationship that doesn't follow a foreign key. I don't know if there's a way to do that properly with a 0..N relationship in Yii.

**Comment by marshl** [comment]
There are some commits in here that shouldn't be. A local rebase might be necessary

**Pull request closed without merging**